File: composeApp/src/commonMain/kotlin/org/sesacteamproject/passmate/ui/payment/CoinHistoryViewModel.kt
package org.sesacteamproject.passmate.ui.payment

import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import org.sesacteamproject.passmate.core.model.onFailure
import org.sesacteamproject.passmate.core.model.onSuccess
import org.sesacteamproject.passmate.mvi.MviViewModel
import org.sesacteamproject.passmate.payment.domain.usecase.GetCoinTransactionsUseCase
import org.sesacteamproject.passmate.payment.domain.usecase.GetMyCoinsUseCase

class CoinHistoryViewModel(
    private val getMyCoinsUseCase: GetMyCoinsUseCase,
    private val getCoinTransactionsUseCase: GetCoinTransactionsUseCase
) : MviViewModel<CoinHistoryUiState, CoinHistoryAction, CoinHistoryEvent>(CoinHistoryUiState()) {

    private var isEntered = false

    private fun onEnter() {
        if (isEntered) {
            return
        }
        isEntered = true
        load()
    }

    // 보유 코인은 실패해도 화면을 덮지 않는다 — 실패 화면은 "목록" 로드 실패일 때만이다 (E-List)
    private suspend fun loadBalance() {
        getMyCoinsUseCase.invoke()
            .onSuccess { coins -> _uiState.update { it.copy(balance = coins.balance) } }
            .onFailure { _uiState.update { it.copy(balance = null) } }
    }

    private suspend fun loadTransactions() {
        getCoinTransactionsUseCase.invoke(cursor = null)
            .onSuccess { page ->
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        hasError = false,
                        items = page.items,
                        hasNext = page.hasNext,
                        nextCursor = page.nextCursor
                    )
                }
            }
            .onFailure { _uiState.update { it.copy(isLoading = false, hasError = true) } }
    }

    private fun load() {
        _uiState.update { it.copy(isLoading = true, hasError = false) }
        viewModelScope.launch { loadBalance() }
        viewModelScope.launch { loadTransactions() }
    }

    private fun onLoadMore() {
        val state = _uiState.value

        if (state.isLoading || state.isLoadingMore || !state.hasNext) {
            return
        }
        _uiState.update { it.copy(isLoadingMore = true) }
        viewModelScope.launch {
            getCoinTransactionsUseCase.invoke(cursor = state.nextCursor)
                .onSuccess { page ->
                    _uiState.update {
                        it.copy(
                            isLoadingMore = false,
                            items = it.items + page.items,
                            hasNext = page.hasNext,
                            nextCursor = page.nextCursor
                        )
                    }
                }
                .onFailure {
                    _uiState.update { it.copy(isLoadingMore = false) }
                    _event.emit(CoinHistoryEvent.ShowNotice("내역을 더 불러오지 못했어요"))
                }
        }
    }

    private fun onSelectFilter(filter: CoinHistoryFilter) {
        _uiState.update { it.copy(filter = filter) }
    }

    private fun onClickCharge() {
        viewModelScope.launch { _event.emit(CoinHistoryEvent.OpenCoinCharge) }
    }

    override fun onAction(action: CoinHistoryAction) {
        when (action) {
            is CoinHistoryAction.Enter -> onEnter()
            is CoinHistoryAction.Retry -> load()
            is CoinHistoryAction.LoadMore -> onLoadMore()
            is CoinHistoryAction.SelectFilter -> onSelectFilter(action.filter)
            is CoinHistoryAction.ClickCharge -> onClickCharge()
        }
    }
}

File: iosApp/iosApp/ui/payment/CoinHistoryViewModel.swift
import Combine
import Foundation
import Shared

// Compose CoinHistoryViewModel.kt 미러 — 보유 코인 + 코인 내역 페이징 로드
final class CoinHistoryViewModel: ObservableObject {
    private let getMyCoinsUseCase: GetMyCoinsUseCase

    private let getCoinTransactionsUseCase: GetCoinTransactionsUseCase

    @Published private(set) var uiState = CoinHistoryUiState()

    let event = PassthroughSubject<CoinHistoryEvent, Never>()

    private var isEntered = false

    private func onEnter() {
        if isEntered {
            return
        }
        isEntered = true
        load()
    }

    // 보유 코인은 실패해도 화면을 덮지 않는다 — 실패 화면은 "목록" 로드 실패일 때만이다 (E-List)
    private func loadBalance() {
        getMyCoinsUseCase.invoke { [weak self] result, error in
            DispatchQueue.main.async {
                guard let self else { return }
                if let coins = (result as? AppResultSuccess<AnyObject>)?.value as? CoinBalance {
                    self.uiState.balance = Int(coins.balance)
                } else {
                    self.uiState.balance = nil
                }
            }
        }
    }

    private func loadTransactions() {
        getCoinTransactionsUseCase.invoke(cursor: nil) { [weak self] result, error in
            DispatchQueue.main.async {
                guard let self else { return }
                self.uiState.isLoading = false
                if let page = (result as? AppResultSuccess<AnyObject>)?.value as? PagedResult<AnyObject> {
                    self.uiState.hasError = false
                    self.uiState.items = self.transactions(page)
                    self.uiState.hasNext = page.hasNext
                    self.uiState.nextCursor = page.nextCursor
                } else {
                    self.uiState.hasError = true
                }
            }
        }
    }

    private func load() {
        uiState.isLoading = true
        uiState.hasError = false
        loadBalance()
        loadTransactions()
    }

    private func onLoadMore() {
        if uiState.isLoading || uiState.isLoadingMore || !uiState.hasNext {
            return
        }
        uiState.isLoadingMore = true
        getCoinTransactionsUseCase.invoke(cursor: uiState.nextCursor) { [weak self] result, error in
            DispatchQueue.main.async {
                guard let self else { return }
                self.uiState.isLoadingMore = false
                if let page = (result as? AppResultSuccess<AnyObject>)?.value as? PagedResult<AnyObject> {
                    self.uiState.items += self.transactions(page)
                    self.uiState.hasNext = page.hasNext
                    self.uiState.nextCursor = page.nextCursor
                } else {
                    self.event.send(.showNotice(message: "내역을 더 불러오지 못했어요"))
                }
            }
        }
    }

    private func onSelectFilter(_ filter: CoinHistoryFilter) {
        uiState.filter = filter
    }

    private func onClickCharge() {
        event.send(.openCoinCharge)
    }

    private func transactions(_ page: PagedResult<AnyObject>) -> [CoinTransaction] {
        page.items.compactMap { $0 as? CoinTransaction }
    }

    func action(_ action: CoinHistoryAction) {
        switch action {
        case .enter:
            onEnter()
        case .retry:
            load()
        case .loadMore:
            onLoadMore()
        case let .selectFilter(filter):
            onSelectFilter(filter)
        case .clickCharge:
            onClickCharge()
        }
    }

    init(getMyCoinsUseCase: GetMyCoinsUseCase, getCoinTransactionsUseCase: GetCoinTransactionsUseCase) {
        self.getMyCoinsUseCase = getMyCoinsUseCase
        self.getCoinTransactionsUseCase = getCoinTransactionsUseCase
    }
}
